### PR TITLE
Dynamic 2022, Part 1 - Redesigning midround rolls (Midrounds happen far more frequently now, and scale their power over time)

### PR DESCRIPTION
**Pull request by Mothblocks**
TODO:
- [x] Bring back forcing midround injections. I made the code surrounding this unnecessarily complicated, can be simplified since midround rolls are equidistant.
- [x] Update midround timer when admins change midround threat
- [x] Document the new variables (and remove the old ones from the README).
- [x] Remove reference to now deleted random event hijacking branch.
- [x] Unit test midround ruleset styles.
- [x] Give rulesets the option for a minimum time, like nuclear assault.
- [x] 💃Playtesting💃

## About The Pull Request
Implements the [new midround roll changes](https://hackmd.io/Don-pCSYQlWdlMdTk6i5dg#Solution1).

### 1. Split midround rulesets into heavy/light impact

First, midround rulesets will be split into light/heavy impact categories. For example, sleeper agent and midround thief would be classified as light impact, while blob/ninja would be classified as heavy impact.

This split will then be used to spawn lighter impact antagonists earlier into the round, and higher impact antagonists later into the round.

- Midrounds before 25 minutes = 100% light impact / 0% heavy impact
- Midrounds from 25-60 minutes = Varying chances, increasing in favor of heavy impact after enough time
- Midrounds from 60+ minutes = 0% light impact / 100% heavy impact

Low impact threat rulesets are guaranteed to roll if there is enough threat. High impact threat rulesets will carry the same % chance as they do now. If a heavy impact threat ruleset cannot be rolled, but there is enough threat, a light impact threat ruleset will roll in its place. In the future, this can potentially be changed into not spawning any ruleset if the station is deemed extremely hostile, or even be tweaked into spawning protagonists like ERTs.

Alongside this, rulesets should be able to determine their minimum time required. For example, nuclear assault has terrible numbers--it's very high cost, and very low weight, which are basically the only variables it can configure. In Dynamic 2022, it should be able to have low cost, still low weight (though maybe nto as much), while only being rollable after 70 minutes or so.

### 2. Midround antagonists now roll as intervals, determined by midround threat

In order to make sure 50 midround threat spawns more antagonists than 10 midround threat, the intervals at which midround threats are spawned will change from 15-45 minutes to increments determined by midround threat. Larger midround threat = more, close intervals, smaller midround threat = less, farther intervals.

Any threat not spent on midrounds will still be carried into latejoins, which as before will still consume midround threat, and are unchanged by this document.

## Why It's Good For The Game

Currently, after a random timespan (15 to 35 minutes on LRP, IIRC, higher on MRP), a percent chance is rolled to see if a midround should spawn. This considers several factors in the round, such as number of people dead, number of antags alive, etc.

This often does a good job of matching the pacing of the round, but too often will a midround not roll when it could have, resulting in a ~30 or so minute wait for the next chance.

Midround chances also vary in their impact on the round drastically. A new sleeper agent affects the round significantly less than a blob, but both will delay the next roll by the same range, and both will be delayed by the same range.

Lastly, having a lot of midround threat has far more diminishing returns than having a lot of roundstart threat, precisely because midrounds are not guaranteed to roll, and might not be very common.

## Changelog

:cl:
balance: Midround rolls will now happen more frequently, with their power scaling up over time.
fix: Fixed midrounds not logging properly.
/:cl:

### CODE REVIEW

**Comment by nianjiilical** [comment]
this overall seems great but my one question is: will the increased chance of midround spawns have the side effect of draining a lot of the threat before any high impacts have a chance to roll?

assuming 50 threat is the average median across rounds (i have no idea if it is or not), will high impact midrounds become noticably rarer because low impact midrounds become more frequent and drain off the threat, and if so is that an intentional part of the design?

**Comment by Mothblocks** [comment]
> will the increased chance of midround spawns have the side effect of draining a lot of the threat before any high impacts have a chance to roll?

With just these changes alone, probably! And yeah, that's a problem. My plan if it does is to lower how much light impact rulesets cost, I think sleeper agent still costs 10, and obsessed probably does too.

However, after changes like that, the equidistant midround points *should* have a good enough amount of threat.

It's possible I need to lower the threat cost of everything across the board? Like, change things like dragon to only be 6-7 threat and such.

**Comment by Colovorat** [comment]
Do we really need that system of points? I mean we can just have 10 points where most antags and events cost 1 point, and something like blob/dragon/nukies will cost 3 or maybe more or less points. Like it seems a lot easier to comprehend and I think it wouldn't be such problem as we have system of low /high impact

**Comment by TheBonded** [comment]
>- Midrounds before 25 minutes = 100% low impact / 0% high impact
>- Midrounds from 25-60 minutes = Varying chances, increasing in favor of high impact after enough time
>- Midrounds from 60+ minutes = 0% low impact / 100% high impact

will this be different for mrp? for example, will the 100% high impact change happen at 90+ minutes instead of 60+, since mrp round generally tend to go on longer?

**Comment by Mothblocks** [comment]
> will this be different for mrp? for example, will the 100% high impact change happen at 90+ minutes instead of 60+, since mrp round generally tend to go on longer?

It *can* be, but won't at first.

MRP rounds *tend* to go on longer, but just like how we want to keep LRP rounds from being 60 minutes and go into 90 minute category, we (maintainers) equally want that for MRP, in the other direction.

**Comment by Mothblocks** [comment]
> Do we really need that system of points? I mean we can just have 10 points where most antags and events cost 1 point, and something like blob/dragon/nukies will cost 3 or maybe more or less points. 

We don't want a dozen different systems of "threat", it makes understanding the system harder.

**Comment by Mothblocks** [comment]
Chose numbers that felt right for midrounds, lowered across the board so that this system makes sense. Everything will be adjusted over time.

File: code/game/gamemodes/dynamic/dynamic_midround_rolling.dm [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+/// Gets the chance for a heavy ruleset midround injection, the dry_run argument is only used for forced injection.
+/datum/game_mode/dynamic/proc/get_heavy_midround_injection_chance(dry_run)
+	var/chance = get_injection_chance(dry_run)
+	var/next_midround_roll = next_midround_injection()
+
+	if (random_event_hijacked != HIJACKED_NOTHING)
+		chance += hijacked_random_event_injection_chance
+
+	var/heavy_coefficient = CLAMP01((next_midround_roll - midround_light_upper_bound) * (midround_heavy_lower_bound - midround_light_upper_bound))
```
Comment (Mothblocks) [inline code/game/gamemodes/dynamic/dynamic_midround_rolling.dm:94]: This lerp is totally fucked.

1. This needs to divide.
2. It doesn't factor in round start time
3. This should just USE LERP

**Comment by RikuTheKiller** [comment]
This right here has single-handedly made me no longer hate you blindly, Mothblocks.

Why:
- It's fucking great. No more having to wait in deadchat for 60 minutes only for all the midround injections to return false.

**Comment by Mothblocks** [comment]
sybil likes me more than anyone else every time i hop on there like 3 people say hi :)

**Comment by Mothblocks** [comment]
After getting frustrated at midround roll chance code I've decided that I'm going to completely gut the awful code meant to pace the round that actually just lowers chances significantly, and make midround % chance based entirely on time (and a few minor contextual nudges, such as more dead than alive lowering chance)

**Comment by SinfulBliss** [comment]
This looks awesome and it's exciting to see dynamic finally getting fixed. 
Are cult and revs considered high impact? If so, does that mean we'll no longer have roundstart cult/revs before the 25 minute mark?

**Comment by Mothblocks** [comment]
Cult and revs are not midrounds and thus are not in scope for this PR in specific.

**Comment by Mothblocks** [comment]
Currently being actively playtested on Sybil.

**Comment by TheSilverNuke** [comment]
Is nuclear assault latejoin or midround? We've been seeing a lot of those recently, they're only like, 7 threat.

**Comment by Mothblocks** [comment]
Midround--they're 7 threat because if you make them too high threat, they'll never roll because they take 70 minutes to spawn.

**Comment by Mothblocks** [comment]
Round IDs would help BTW

**Comment by Mothblocks** [comment]
Need to collect data on frequencies of rulesets with the TM active.

**Comment by Mothblocks** [comment]
And then also run on Manuel

**Comment by Mothblocks** [comment]
Actually I'm just going to enable it on Manuel and see what happens

**Comment by Annihilit** [comment]
I know this would require a sprite and stuff, but it'd be really cool if the artifact spawned as a physical obelisk somewhere in science/security/command.

"A massive stone monolith of unknown origin. Crimson light leaks through the cracks in its carved skin, and when you put your ear to it you hear an inaudible cry from within"
Would be cooler than just the command report. Also a great WTF moment for players who don't know what it is

**Comment by Mothblocks** [comment]
Going to try TMing on Manuel again

**Comment by TemporalOroboros** [comment]
> I know this would require a sprite and stuff, but it'd be really cool if the artifact spawned as a physical obelisk somewhere in science/security/command.
> 
> "A massive stone monolith of unknown origin. Crimson light leaks through the cracks in its carved skin, and when you put your ear to it you hear an inaudible cry from within" Would be cooler than just the command report. Also a great WTF moment for players who don't know what it is

I think this is supposed to be in reference to #66034, which is... not here.

**Comment by Mothblocks** [comment]
https://tgstation13.org/parsed-logs/manuel/data/logs/2022/05/04/round-182574/dynamic.json

Nuclear assault running before the expected timeframe

**Comment by Mothblocks** [comment]
Fun bug! Only blob, blob infection, nuclear assault, and alien infestation were running, and only before when they should

**Comment by MrMelbert** [comment]
Midround wiz seems common now - I observed 3 rounds in sequence on manuel - 183001 (48 threat), 183004 (70 threat), 183008 (48 threat, cancelled it, rolled again later anyways), all of them rolled midround wiz

**Comment by Mothblocks** [comment]
Wizard already has the lowest weight you can have so I'm not quite sure what else to do there assuming the code is correct, if cost is raised then it just straight up will not run because the point at which heavy rulesets run is enough to where there's fleeting amounts of threat left

I will take a look at rates

**Comment by Annihilit** [comment]
> > I know this would require a sprite and stuff, but it'd be really cool if the artifact spawned as a physical obelisk somewhere in science/security/command.
> > "A massive stone monolith of unknown origin. Crimson light leaks through the cracks in its carved skin, and when you put your ear to it you hear an inaudible cry from within" Would be cooler than just the command report. Also a great WTF moment for players who don't know what it is
> 
> I think this is supposed to be in reference to #66034, which is... not here.

I genuinely have no idea how it ended up here

**Comment by Mothblocks** [comment]
Need to finish up docs, and then also I want to implement threat level being influenced by *very low* (sub-20) populations, currently low pop will constantly throw traitors because Dynamic 2022 forces all midrounds to execute.

**Comment by MetalClone** [comment]
MRP server is not ready for this change, the culture on Manuel is already having a decline in security and command players. None of them are interested in being the sole-defender of the station versus the constant antagonists pouring in from the midround rolls. The Station is often irreparably destroyed well before the shuttle is called at this point due to the constant ninjas, space dragons, and wizards. At the very least give antag rolls who can summon other antags (demons, carp armies, revs, and cultists) a higher threat level.

**Comment by Mothblocks** [comment]
I would disagree that Manuel is not ready for this--I am observing 1 hour 40 rounds from Manuel with Dynamic 2022 on. That is, in part, the goal.

Revs and cultists are not in scope of this change, which only factors in midrounds. 

Wizard has already been made rarer, even considering it's already the *lowest weight of any ruleset*.

Higher threat cost does not mean much because the big change here is equidistant midround rolls. Increasing threat cost of something will mostly mean that other rulesets *can't* run, rather than make it rarer.

**Comment by tralezab** [comment]
ooops i clicked the wrong btton ssry guys

**Comment by refticus** [comment]
this test merge has been nothing but a problem on manuel during lowpop hours, with crews being almost entirely composed of traitors by the end of a round.

for reference, [round id 183399](https://sb.atlantaned.space/rounds/183399), 14 total pop, 6 traitors, about 4 of the crew members were killed or went SSD. the end of the round was a stalemate with the traitors teaming up and doing little other than bug objectives until they got bored.

i understand the purpose of this change, but there really needs to be a minimum pop count before the effects start to kick in. traitors are meant to be scarce and hard to identify, not the entire crew where they end up just teaming together to slowly accumulate TC for end of the round gags.

**Comment by SinfulBliss** [comment]
> this test merge has been nothing but a problem on manuel during lowpop hours, with crews being almost entirely composed of traitors by the end of a round.
> 
> for reference, [round id 183399](https://sb.atlantaned.space/rounds/183399), 14 total pop, 6 traitors, about 4 of the crew members were killed or went SSD. the end of the round was a stalemate with the traitors teaming up and doing little other than bug objectives until they got bored.
> 
> i understand the purpose of this change, but there really needs to be a minimum pop count before the effects start to kick in. traitors are meant to be scarce and hard to identify, not the entire crew where they end up just teaming together to slowly accumulate TC for end of the round gags.

As someone who has played a ton of lowpop, it's nice having more antagonists. 14pop is definitely not the ideal amount of players for a round and without traitors it can be extremely slow, I would consider that a worthy tradeoff. If less than 4 of the 16 were killed it doesn't seem like a big deal, although I can see it being annoying since Manuel forbids validhunting (so you just have to sit there while 6 traitors do their dirt).

**Comment by refticus** [comment]
> As someone who has played a ton of lowpop, it's nice having more antagonists. 14pop is definitely not the ideal amount of players for a round and without traitors it can be extremely slow, I would consider that a worthy tradeoff. If less than 4 of the 16 were killed it doesn't seem like a big deal, although I can see it being annoying since Manuel forbids validhunting (so you just have to sit there while 6 traitors do their dirt).

that's exactly the problem however, part of lowpop and MRP on manuel is that the rounds aren't just a flash, bombing medbay within the first half hour of a round is not the norm as traitors prefer to be more nuanced and stealthy with their actions.

however, what's the point of that approach when at least half of the crew you'll meet are also traitors and any semblance of resistance has been dealt with long before you appeared? who are you meant to antagonise against when there's nobody but you, two hermits doing their job, the clown, a borg, and 5 other traitors?

the amount of traitors/antagonists to normal crew is something that needs to be balanced incredibly carefully; a lack of active antagonists leads to nothing happening, a majority of antagonists leads to the antagonists getting bored because there's nobody to antagonise.

this PR has, from experience, only produced rounds of the latter and it does this in a consistent and predictable way. from what i can tell, this PR is more designed around terry and its mid-high pop culture of validhunting as you'd need a constant flow of traitors to keep up with those that end up executed or permabrigged, but on a MRP server with a different and slower culture this change only harms the experience and solves _nothing_

**Comment by Mothblocks** [comment]
> for reference, [round id 183399](https://sb.atlantaned.space/rounds/183399), 14 total pop, 6 traitors, about 4 of the crew members were killed or went SSD. the end of the round was a stalemate with the traitors teaming up and doing little other than bug objectives until they got bored.

This has already been fixed in https://github.com/tgstation/tgstation/pull/65997/commits/7d029ad19417f392428f18f7dc56206c41067b5a, the test merge was just not updated.

**Pull request merged**